server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -347,67 +341,17 @@ private synchronized DocumentMapper internalMerge(DocumentMapper mapper, MergeRe
-        Collections.addAll(fieldMappers, metadataMappers);
-        MapperUtils.collect(newMapper.mapping().root(), objectMappers, fieldMappers, fieldAliasMappers);
-
-        MapperMergeValidator.validateNewMappers(objectMappers, fieldMappers, fieldAliasMappers);
-        checkPartitionedIndexConstraints(newMapper);
-
-        // update lookup data-structures
-        FieldTypeLookup newFieldTypes = new FieldTypeLookup(fieldMappers, fieldAliasMappers);
-
-        for (ObjectMapper objectMapper : objectMappers) {
-            if (fullPathObjectMappers == this.fullPathObjectMappers) {
-                // first time through the loops
-                fullPathObjectMappers = new HashMap<>(this.fullPathObjectMappers);
-            }
-            fullPathObjectMappers.put(objectMapper.fullPath(), objectMapper);
-
-            if (objectMapper.nested().isNested()) {
-                hasNested = true;
-            }
-        }
-
-        MapperMergeValidator.validateFieldReferences(fieldMappers, fieldAliasMappers,
-            fullPathObjectMappers, newFieldTypes, metadataMappers, newMapper);
-
-        ContextMapping.validateContextPaths(indexSettings.getIndexVersionCreated(), fieldMappers, newFieldTypes::get);
-
-        if (reason != MergeReason.MAPPING_RECOVERY) {
-            // These checks will only be performed on the master node when an index is created, or
-            // there is a call to the update mapping API. For all other cases like the master node
-            // restoring mappings from disk or data nodes deserializing cluster state that was sent
-            // by the master node, these checks will be skipped.
-            // Also, don't take metadata mappers into account for the field limit check
-            checkTotalFieldsLimit(objectMappers.size() + fieldMappers.size() - metadataMappers.length
-                + fieldAliasMappers.size() );
-            checkFieldNameSoftLimit(objectMappers, fieldMappers, fieldAliasMappers);
-            checkNestedFieldsLimit(fullPathObjectMappers);
-            checkDepthLimit(fullPathObjectMappers.keySet());
-        }
-        checkIndexSortCompatibility(indexSettings.getIndexSortConfig(), hasNested);
+        newMapper.validate(indexSettings, reason != MergeReason.MAPPING_RECOVERY);
 
         if (reason == MergeReason.MAPPING_UPDATE_PREFLIGHT) {
             return newMapper;
         }
 
-        // only need to immutably rewrap these if the previous reference was changed.
-        // if not then they are already implicitly immutable.
-        if (fullPathObjectMappers != this.fullPathObjectMappers) {
-            fullPathObjectMappers = Collections.unmodifiableMap(fullPathObjectMappers);
-        }
-
         // commit the change
         this.mapper = newMapper;
-        this.fieldTypes = newFieldTypes;
-        this.hasNested = hasNested;
-        this.fullPathObjectMappers = fullPathObjectMappers;
+        this.fieldTypes = newMapper.fieldTypes();
```

Comment:
Instead of storing these data structures again, should `MapperService` just delegate to `mapper` to retrieve field types, etc. That would clarify who actually 'owns' this data.

Reply:
++